src/main/java/org/manifold/compiler/front/FunctionValueVertex.java

```diff
@@ -0,0 +1,158 @@
+    for (String argName : inputType.getSubtypes().keySet()) {
+      ExpressionEdge e = new ExpressionEdge(null, null); // I know what I'm doing
+      functionBody.addEdge(e);
+      inputEdges.put(argName, e);
+    }
+    for (String argName : outputType.getSubtypes().keySet()) {
+      // "name resolution": look for a variable reference vertex with this name in the subgraph
+      for (Map.Entry<VariableIdentifier, VariableReferenceVertex> vRef
+          : functionBody.getVariableVertices().entrySet()) {
+        if (vRef.getKey().getName().equals(argName)) {
+          ExpressionEdge e = new ExpressionEdge(vRef.getValue(), null);
+          functionBody.addEdge(e);
+          outputEdges.put(argName, e);
+          break;
+        }
+      }
+    }
+    TupleValueVertex vInput = new TupleValueVertex(functionBody, inputEdges);
+    functionBody.addVertex(vInput);
+    TupleValueVertex vOutput = new TupleValueVertex(functionBody, outputEdges);
+    functionBody.addVertex(vOutput);
+    function = new FunctionValue(type, functionBody, vInput, vOutput);
+    // inject StaticAttributeAccesses to assign each variable from the input
+    // TODO we're making a few guesses about "name resolution" here
+    for (Map.Entry<VariableIdentifier, VariableReferenceVertex> entry
+          : functionBody.getVariableVertices().entrySet()) {
+      String varName = entry.getKey().getName();
+      if (inputType.getSubtypes().containsKey(varName)) {
+        ExpressionEdge eInput = new ExpressionEdge(vInput, null);
+        functionBody.addEdge(eInput);
+        StaticAttributeAccessVertex vAccess = new StaticAttributeAccessVertex(
+            functionBody, eInput, varName);
+        functionBody.addVertex(vAccess);
+        ExpressionEdge eAssign = new ExpressionEdge(vAccess, entry.getValue());
+        functionBody.addEdge(eAssign);
+      }
+    }
+  }
+
+  @Override
+  public void verify() throws Exception {
+    // TODO Auto-generated method stub
+
+  }
+
+  @Override
+  public boolean isElaborationtimeKnowable() {
+    return true;
+  }
+
+  @Override
+  public boolean isRuntimeKnowable() {
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    if (function == null) {
+      return "function value (not elaborated)";
+    } else {
+      return "function value " + type.toString();
+    }
+  }
+
+  @Override
+  public void writeToDOTFile(BufferedWriter writer) throws IOException {
+    String objectID = Integer.toString(System.identityHashCode(this));
+    String label = this.toString();
+    writer.write(objectID);
+    writer.write(" [");
+    writer.write("label=\"");
+    writer.write(objectID);
+    writer.write("\n");
+    writer.write(label);
+    writer.write("\"");
+    writer.write("];");
+    writer.newLine();
+  }
+
+  @Override
+  public ExpressionVertex copy(ExpressionGraph g,
+      Map<ExpressionEdge, ExpressionEdge> edgeMap) {
+    // TODO Auto-generated method stub
```

Comment:
Right, and this function is required to create the copied instance of FunctionInvocationVertex.